pkg/providers/gke.go

```diff
@@ -0,0 +1,342 @@
+	// update internal cache
+	if lb, ok := i.crToLB[crName]; ok {
+		lbName = lb.Name
+		delete(i.crToLB, crName)
+		delete(i.lbToCRs[lb], crName)
+		for _, svcPort := range clusterSvc.Spec.Ports {
+			delete(i.lbToPorts[lb], svcPort.Port)
+		}
+		log.WithName("GKE").Info("DeassociateLB", "cr", crName, "lb", lb)
+	}
+	firewallName := lbName+"-fwrule"+fmt.Sprint(clusterSvc.Spec.Ports[0].Port)
+	i.FirewallDelete(firewallName)	
+	return nil
+}
+
+func (i *GKE) UpdateService(svc, lb *corev1.Service) (bool, bool) {
+	lbName := types.NamespacedName{Name: lb.Name, Namespace: lb.Namespace}
+	occupiedPorts := i.lbToPorts[lbName]
+	if len(occupiedPorts) == 0 {
+		occupiedPorts = int32Set{}
+	}
+	portUpdated := updatePort(svc, lb, occupiedPorts)
+	return portUpdated, true
+}
+
+func (g *GKE) createStaticIP(name, ip string) {
+   addr := &compute.Address {
+             Name: name,
+             Address: ip,
+             Region: g.region,
+             }
+   addrInsertCall := g.client.Addresses.Insert(g.project,g.region,addr)
+   _, err := addrInsertCall.Do()
+
+    if err != nil {
+        log.WithName("gke").Error(err, "Faied to secure a Static IP")
+    }
+}
+
+func (g *GKE) getTargetPool_usingLBName(lb string)  *compute.TargetPool {
+    resp, err := g.client.TargetPools.List(g.project,g.region).Context(ctx).Do()
+    if err != nil {
+        log.WithName("gke").Error(err,"Failed to find the loadbalancer "+lb)
+    }
+    for _,item := range resp.Items {
+        if strings.Contains(item.Description, lb) {
+                return item;
+        }
+    }
+    return nil
+}
+
+func (g *GKE) GetForwardingRule(name string) *compute.ForwardingRule {
+    resp, err := g.client.ForwardingRules.List(g.project,g.region).Context(ctx).Do()
+    if err != nil {
+	    	if strings.Contains(err.Error(), "Error 404") {
+	    		return nil
+	    	}
+        log.WithName("GKE").Error(err, "Failed to get Forwarding rules")
+    }
+    for _,item := range resp.Items {
+        if strings.Compare(item.Name, name)==0 {
+                return item
+        }
+    }
+    return nil
+}
+
+func (g *GKE) ForwardingRuleInsert(name, ip, item string) {
+   fw := &compute.ForwardingRule {
+		   	IPAddress: ip,
+             IPProtocol: "TCP",
+             Description: "KubeConDemo, forwarding rule to reach service through port ",
+             LoadBalancingScheme: "EXTERNAL",
+             Target: item,
+             Name : name,
+             PortRange: "28000-33000",
+             }
+   fwInsertCall := g.client.ForwardingRules.Insert(g.project,g.region, fw)
+   _, err := fwInsertCall.Do()
+
+    if err != nil {
+        log.WithName("gke").Error(err, "Failed to create the Forwarding Rule for the item "+item)
+    }
+}
+
+func (g *GKE) ForwardingRuleDelete(name string) {
+   fwDeleteCall := g.client.ForwardingRules.Delete(g.project,g.region,name)
+   _, err := fwDeleteCall.Do()
+
+    if err != nil {
+        log.WithName("gke").Error(err, "Failed to delete the firewall rule "+name)
```

Comment:
Please return potential error.

Reply:
Yes, error handling must percolate up, but in the next iteration.